### PR TITLE
add two telemetry analyzers

### PR DESCRIPTION
**Pull request by zengyuxing007**
xref: https://github.com/istio/istio/issues/43705

@zirain @howardjohn  @hzxuzhonghu

### CODE REVIEW

File: pkg/config/analysis/analyzers/telemetry/selector.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package telemetry
+
+import (
+	"fmt"
+	"istio.io/api/telemetry/v1alpha1"
```
Comment (hanxiaop) [inline pkg/config/analysis/analyzers/telemetry/selector.go:18]: format

File: pkg/config/analysis/analyzers/telemetry/selector.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+// Metadata implements Analyzer
+func (a *SelectorAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name: "telemetry.SelectorAnalyzer",
+		Description: "Validates that telemetries that define a selector " +
+			"match at least one pod, and that there aren't multiple telemetry resources that select overlapping pods",
+		Inputs: []config.GroupVersionKind{
+			gvk.Telemetry,
+			gvk.Pod,
```
Comment (hanxiaop) [inline pkg/config/analysis/analyzers/telemetry/selector.go:45]: What about the deployment resource? If we are analyzing file resources, currently we won't have pods.

File: releasenotes/notes/43706.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+apiVersion: release-notes/v2
+kind: feature
+area: installation
```
Comment (hanxiaop) [inline releasenotes/notes/43706.yaml:3]: This should be `istioctl`

File: releasenotes/notes/43706.yaml [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
+issue:
+- 43705
+
+releaseNotes:
+- |
+  **Added** analyzer for telemetry resource.
```
Comment (hanxiaop) [inline releasenotes/notes/43706.yaml:10]: Include more details about what is being analyzed?

Comment (zengyuxing007) [reply releasenotes/notes/43706.yaml]: Done

File: releasenotes/notes/43706.yaml (hunk 3 above)
Comment (zengyuxing007) [reply releasenotes/notes/43706.yaml]: Done

File: pkg/config/analysis/analyzers/telemetry/selector.go (hunk 1 above)
Comment (zengyuxing007) [reply pkg/config/analysis/analyzers/telemetry/selector.go]: Done

File: pkg/config/analysis/analyzers/telemetry/selector.go (hunk 2 above)
Comment (zengyuxing007) [reply pkg/config/analysis/analyzers/telemetry/selector.go:45]: This is to analyze whether the selector has selected a valid pod

Comment (hanxiaop) [reply pkg/config/analysis/analyzers/telemetry/selector.go:45]: This works for in-cluster analysis. However, we should also consider the option of running analyze against file yaml configs, there won't be pod resources.

Comment (zengyuxing007) [reply pkg/config/analysis/analyzers/telemetry/selector.go:45]: Sorry, didn't quite understand. The yaml file is added under test case (testdata/telemetry-selector.yaml) and you can see the pod related yaml configured under this file,
Is this what you mean?

Comment (hanxiaop) [reply pkg/config/analysis/analyzers/telemetry/selector.go:45]: Yes. However, for file configs, we won’t directly use pods. Instead, we will use some high-level abstraction like deployments. Currently, when you try to run analyze against files, the analyzers won’t convert deployments to pods. This is different from the in-cluster environment. As a result, the deployments from files won’t get analyzed at all.

File: pkg/config/analysis/analyzers/testdata/telemetry-selector.yaml [hunk 5]
```diff
@@ -0,0 +1,123 @@
+apiVersion: v1
+kind: Pod
```
Comment (hanxiaop) [inline pkg/config/analysis/analyzers/testdata/telemetry-selector.yaml:2]: Revise to Deployment

File: pkg/config/analysis/analyzers/telemetry/selector.go (hunk 2 above)
Comment (zengyuxing007) [reply pkg/config/analysis/analyzers/telemetry/selector.go:45]: Thank you for your patience in replying。 it is done.

**Review by hzxuzhonghu — APPROVED** [review]

File: releasenotes/notes/43706.yaml [hunk 6]
```diff
@@ -0,0 +1,14 @@
+apiVersion: release-notes/v2
+kind: feature
+area: istioctl
+
+issue:
+- 43705
+
+releaseNotes:
+- |
+  **Added** `istioctl analyzer` will display a error when encountering the following two situations
```
Comment (ericvn) [inline releasenotes/notes/43706.yaml:10]:
```suggestion
  **Added** `istioctl analyzer` will display a error when encountering the following two situations:
```

File: releasenotes/notes/43706.yaml [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+kind: feature
+area: istioctl
+
+issue:
+- 43705
+
+releaseNotes:
+- |
+  **Added** `istioctl analyzer` will display a error when encountering the following two situations  
+  - For any namespace,if there are multiple Telemetry CRs without selector
```
Comment (ericvn) [inline releasenotes/notes/43706.yaml:11]:
```suggestion
  - For any namespace, if there are multiple Telemetry CRs without selector,
```

File: releasenotes/notes/43706.yaml [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+area: istioctl
+
+issue:
+- 43705
+
+releaseNotes:
+- |
+  **Added** `istioctl analyzer` will display a error when encountering the following two situations  
+  - For any namespace,if there are multiple Telemetry CRs without selector 
+  - For any namespace,if there are multiple Telemetry CRs with selector which selected for the same workload
```
Comment (ericvn) [inline releasenotes/notes/43706.yaml:12]:
```suggestion
  - For any namespace, if there are multiple Telemetry CRs with selector which are selected for the same workload.
```

Comment (zengyuxing007) [reply releasenotes/notes/43706.yaml]: thanks, done

File: releasenotes/notes/43706.yaml (hunk 7 above)
Comment (zengyuxing007) [reply releasenotes/notes/43706.yaml]: Done

File: releasenotes/notes/43706.yaml (hunk 6 above)
Comment (zengyuxing007) [reply releasenotes/notes/43706.yaml]: Done

File: pkg/config/analysis/analyzers/telemetry/selector.go (hunk 2 above)
Comment (howardjohn) [reply pkg/config/analysis/analyzers/telemetry/selector.go:45]: see https://github.com/istio/istio/pull/43712#pullrequestreview-1322119145. I would have this PR go with the current mechanism, and if we change it we can change all analyzers.

File: pkg/config/analysis/msg/messages.yaml [hunk 9]
```diff
@@ -618,3 +618,29 @@ messages:
     args:
       - name: podNames
         type: "[]string"
+
+  - name: "ConflictingTelemetryWorkloadSelectors"
```
[the remark this answers is not in the document]
Comment (zengyuxing007) [reply pkg/config/analysis/msg/messages.yaml:622]: Yes, Gateway is actually similar， probably later we can write a generic。

[the remark this answers is not in the document]
Comment (zengyuxing007) [reply pkg/config/analysis/msg/messages.yaml:622]: Now, this piece is really not perfect at the moment, but similar to analyzing the local yaml file (#43712), maybe we can refactor it together later when we think it through.

**Review by zirain — APPROVED** [review]
Generally looks good, I think we can follow up [john's comment](https://github.com/istio/istio/pull/43706#discussion_r1123334819) in a separate PR.

File: releasenotes/notes/43706.yaml [hunk 10]
```diff
@@ -0,0 +1,14 @@
+apiVersion: release-notes/v2
+kind: feature
+area: istioctl
+
+issue:
+- 43705
+
+releaseNotes:
+- |
+  **Added** `istioctl analyzer` will display a error when encountering the following two situations:
```
Comment (hanxiaop) [inline releasenotes/notes/43706.yaml:10]:
```suggestion
  **Added** `istioctl analyze` will display a error when encountering the following two situations:
```

**Review by hanxiaop — APPROVED** [review]

**Pull request merged**